platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/TypedResolveException.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.ivyservice;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.artifacts.ResolveException;
+
+import java.util.List;
+
+/**
+ * An internal specialization of the public {@link ResolveException}. All resolve exceptions thrown
+ * by Gradle are assumed to be an instance of this class.
+ */
+public class TypedResolveException extends ResolveException {
+    private final String type;
+    private final ImmutableList<String> resolutions;
+
+    /**
+     * Creates a new instance without resolutions.
+     */
+    public TypedResolveException(String type, String displayName, Iterable<? extends Throwable> failures) {
+        super(buildMessage(type, displayName), failures, false);
+        this.type = type;
+        this.resolutions = ImmutableList.of();
+    }
+
+    /**
+     * Creates a new instance with resolutions.
+     */
+    public TypedResolveException(String type, String displayName, Iterable<? extends Throwable> failures, List<String> resolutions) {
+        super(buildMessage(type, displayName), failures, false);
+        this.type = type;
+        this.resolutions = ImmutableList.copyOf(resolutions);
+    }
+
+    /**
+     * The type. This is used to build the failure message.
+     *
+     * Usually "dependencies", "artifacts", or "files".
+     */
+    public String getType() {
+        return type;
+    }
+
+    @Override
+    public List<String> getResolutions() {
+        return ImmutableList.<String>builder()
```

Comment:
In practice, `initCause` and `initCauses` should only be called immediately after constructing this exception. IMO we should have `initCause` throw an exception and delete `initCauses`, and force the causes to be supplied upon construction. 

However, this touches a little too many classes than I would like to add for this PR.

In terms of `getResolutions` assuming we made the above changes I think we _could_ calculate the resolutions upon-construction, but I don't necessarily think we _should_ do that. 

I don't think that its a problem that we return different instances of this list for each call to `getResoultions`. The method is called by `BuildExceptionReporter` when this error bubbles-up and is printed to the console, so we definitely need to properly implement this method if we want the user to see these resolutions.

Reply:
Overriding the init methods to throw would also not work (without a little massaging at least), since the super constructor we call calls `initCauses`

Reply:
> assuming we made the above changes I think we could calculate the resolutions upon-construction, but I don't necessarily think we should do that.

😮 Wasn't expecting you to be the one advocating for mutability.  It's not only different instances if you call it at different times, it's potentially different contents...at which point, why bother returning an `ImmutableList` at all?  It gives a false impression.

Making `DefaultMultiCauseException` require causes upon construction and getting rid of `initCause(s)` outside of that seems like maintenance work that we should do at some point, but yeah, it's outside the scope here.

If we're going to leave this, can it just return a plain old `ArrayList` and add an `@implNote` about what it's doing?  I'd consider this "unexpected behavior" for something that appears a simple getter.